src/ParaTest/Runners/PHPUnit/SuiteLoader.php

```diff
@@ -214,29 +214,96 @@ private function executableTests($path, $classMethods)
     /**
      * Identify method dependencies, and group dependents and dependees on a single methodBatch
      * If no dependencies exist each methodBatch will contain a single method.
-     * @param  array of ParsedFunction $classMethods
+     * @param  string      $path
+     * @param  ParsedClass $class
      * @return array of MethodBatches. Each MethodBatch has an array of method names
      */
-    private function getMethodBatches($classMethods)
+    private function getMethodBatches($path, $class)
     {
+        $classMethods = $class->getMethods($this->options ? $this->options->annotations : array());
+        $maxBatchSize = $this->options && $this->options->functional ? $this->options->maxBatchSize : 0;
         $methodBatches = array();
         foreach ($classMethods as $method) {
+            $tests = $this->getMethodTests($path, $class, $method);
+            if (!$tests) {
+                continue;
+            }
+
             if (($dependsOn = $this->methodDependency($method)) != null) {
                 foreach ($methodBatches as $key => $methodBatch) {
                     foreach ($methodBatch as $methodName) {
                         if ($dependsOn === $methodName) {
-                            $methodBatches[$key][] = $method->getName();
+                            $methodBatches[$key] = array_merge($methodBatches[$key], $tests);
                             continue;
                         }
                     }
                 }
             } else {
-                $methodBatches[] = array($method->getName());
+                foreach ($tests as $test) {
+                    $lastIndex = count($methodBatches) - 1;
+                    if ($lastIndex != -1
+                        && count($methodBatches[$lastIndex]) < $maxBatchSize
+                    ) {
+                        $methodBatches[$lastIndex][] = $test;
+                    } else {
+                        $methodBatches[] = array($test);
+                    }
+                }
             }
         }
         return $methodBatches;
     }
 
+    private function getMethodTests($path, $class, $method)
+    {
+        $result = array();
+
+        $dataProvider = $this->methodDataProvider($method);
+        if ($dataProvider) {
```

Comment:
Would be nice if this branch could be covered on unit level. It's covered on functional level but based on the complexity of these methods, a unittest would fit in here very well.